c2/c2-client-bundle/c2-client-http/src/test/java/org/apache/nifi/c2/client/http/C2HttpClientTest.java

```diff
@@ -0,0 +1,158 @@
+
+import com.github.tomakehurst.wiremock.client.WireMock;
+import com.github.tomakehurst.wiremock.junit5.WireMockRuntimeInfo;
+import com.github.tomakehurst.wiremock.junit5.WireMockTest;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import org.apache.nifi.c2.client.C2ClientConfig;
+import org.apache.nifi.c2.client.api.C2Serializer;
+import org.apache.nifi.c2.protocol.api.C2Heartbeat;
+import org.apache.nifi.c2.protocol.api.C2HeartbeatResponse;
+import org.apache.nifi.c2.protocol.api.C2OperationAck;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@WireMockTest
+@ExtendWith(MockitoExtension.class)
+public class C2HttpClientTest {
+
+    private static final String HEARTBEAT_PATH = "/c2/heartbeat";
+    private static final String UPDATE_PATH = "/c2/update";
+    private static final String ACK_PATH = "/c2/acknowledge";
+    private static final int HTTP_STATUS_OK = 200;
+
+    @Mock
+    private C2ClientConfig c2ClientConfig;
+
+    @Mock
+    private C2Serializer serializer;
+
+    @InjectMocks
+    private C2HttpClient c2HttpClient;
+
+    @Test
+    void shouldReturnC2HeartbeatResponse(WireMockRuntimeInfo wmRuntimeInfo) {
+        // given
+        C2HeartbeatResponse hbResponse = new C2HeartbeatResponse();
+        stubFor(WireMock.post(HEARTBEAT_PATH)
+            .willReturn(aResponse().withBody("dummyResponseBody")));
+
+        given(serializer.serialize(any(C2Heartbeat.class))).willReturn(Optional.of("Heartbeat"));
+        given(serializer.deserialize(any(), any())).willReturn(Optional.of(hbResponse));
+        given(c2ClientConfig.getC2Url()).willReturn(wmRuntimeInfo.getHttpBaseUrl() + HEARTBEAT_PATH);
+
+        // when
+        Optional<C2HeartbeatResponse> response = c2HttpClient.publishHeartbeat(new C2Heartbeat());
+
+        // then
+        assertTrue(response.isPresent());
+        assertEquals(response.get(), hbResponse);
+    }
+
+    @Test
+    void shouldReturnEmptyInCaseOfCommunicationIssue() {
+        // given
+        given(serializer.serialize(any(C2Heartbeat.class))).willReturn(Optional.of("Heartbeat"));
+        given(c2ClientConfig.getC2Url()).willReturn("http://localhost/incorrectPath");
+
+        // when
+        Optional<C2HeartbeatResponse> response = c2HttpClient.publishHeartbeat(new C2Heartbeat());
+
+        // then
+        assertFalse(response.isPresent());
+    }
+
+    @Test
+    void shouldThrowExceptionForInvalidKeystoreFilenameAtInitialization() {
+        // given
+        given(c2ClientConfig.getKeystoreFilename()).willReturn("incorrectKeystoreFilename");
+
+        // when
+        IllegalStateException exception = assertThrows(IllegalStateException.class, () -> new C2HttpClient(c2ClientConfig, serializer));
+
+        // then
+        assertEquals("OkHttp TLS configuration failed", exception.getMessage());
```

Comment:
In general, it is best to avoid test an exact exception message, since that should not be part of the method contract. If the purpose is to test for a specific scenario, recommend either testing that the message contains a keyword, such as `TLS`, or using a different exception type.